[13.0][ADD] New module 'stock_helper_delivery'

sebalix opened the pull request:
Include a new helper method to get the first picking with a carrier from an operation in the middle of a chain.
Used for instance to get the carrier from any non-delivery operation, it will returns the related delivery which has a carrier defined.

sebalix committed:
New module 'stock_helper_delivery'

Include a new helper method to get the carrier from a operation in the
middle of a chain.
Used for instance to get the carrier from a Pack operation, it will
returns the carrier of the related Delivery operation.

simahawk commented on stock_helper_delivery/__manifest__.py:
```diff
@@ -0,0 +1,15 @@
+    "author": "Camptocamp, Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/stock-logistics-warehouse",
+    "category": "Hidden",
+    "depends": ["stock", "delivery", "stock_helper"],
```
should it really depend on stock_helper?

sebalix replied:
Not really, I was re-using the tests class from `stock_helper` but we could remove this dependency if it's really needed

sebalix replied:
Updated to get rid of this dependency + increase test coverage

sebalix committed:
fixup! New module 'stock_helper_delivery'

remove the dependency on 'stock_helper' and increase test coverage

simahawk commented on stock_helper_delivery/models/stock_picking.py:
```diff
@@ -0,0 +1,32 @@
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def get_carrier_from_chained_picking(self):
```
as reported on the other PR: make it return the picking instead. To be renamed as `get_picking_with_carrier_from_chain` or something like that.

simahawk commented on stock_helper_delivery/models/stock_picking.py:
```diff
@@ -0,0 +1,32 @@
+        if self.carrier_id:
+            return self.carrier_id
+
+        def get_moves_dest(moves):
```
as reported on the other PR: loop and stop at the 1st one that matches.

sebalix committed:
fixup! fixup! New module 'stock_helper_delivery'

change helper method to return the first picking with a carrier instead of the carrier itself

mmequignon commented on stock_helper_delivery/models/stock_picking.py:
```diff
@@ -0,0 +1,31 @@
+                    return move
+
+        move_dest = get_first_move_dest(self.move_lines)
+        while move_dest:
+            if move_dest.picking_id.carrier_id:
+                return move_dest.picking_id
```
```suggestion
        if move_dest:
            if move_dest.picking_id.carrier_id:
                return move_dest.picking_id
```
At most you'll have 1 record. No need to use "while" here, or am I missing something?

simahawk replied:
it means it jumps on the next move till you get a dest move. It stops either if you find the carrier or if you don't have any dest move

mmequignon replied:
Well, if I read get_first_move_dest, it'll return a single record.
What I understand from your explaination is something like this
```python
        def get_first_move_dest(moves):
            for move in moves.move_dest_ids:
                if move.state not in ("cancel", "done"):
                    yield move
                    
        while get_first_move_dest(self.move_lines):
            if move_dest.picking_id.carrier_id:
                return move_dest.picking_id
```
Anyway, I'm probably missing something.
Forget about this

mmequignon reviewed (approved):
technical review : LG

sebalix committed:
fixup! fixup! fixup! New module 'stock_helper_delivery'

'stock_helper_delivery' method to return the related delivery of any operation

TDu commented on stock_helper_delivery/models/stock_picking.py:
```diff
@@ -17,15 +24,21 @@ def get_picking_with_carrier_from_chain(self):
+                return picking
+            move_dest = get_first_move_dest(move_dest, done=done)
+        # Should return an empty record if we reach this line
+        return self.browse()
```
Should it not return an empty move instead of stock.picking ?

sebalix replied:
No this method returns a `stock.picking` record (the delivery one if we found one), it never returns a move.

TDu replied:
:+1:

sebalix commented:
I want to replace this module by a more functional one (not a "helper" one, difficult to maintain depending on the dependencies, and a bit hackish), the new module will implement the same kind of methods usable by other modules.

sebalix commented:
Closing, replaced by https://github.com/OCA/delivery-carrier/pull/354
